File: LeetCode/C/22-backtracking-括号生成.cpp
/*
 * @lc app=leetcode.cn id=22 lang=cpp
 *
 * [22] 括号生成
 */
#include "main.h"

// @lc code=start
class Solution {
  public:
    vector<string> generateParenthesis(int n) {
        if (n != 0)
            backtracking("", n, n);
        return res;
    }

  private:
    vector<string> res;

    void backtracking(string s, int l_rest, int r_rest) {
        if (l_rest == 0 && r_rest == 0) {
            res.push_back(s);
            return;
        }
        if (l_rest > r_rest || l_rest < 0 || r_rest < 0)
            return;
        backtracking(s + '(', l_rest - 1, r_rest);
        backtracking(s + ')', l_rest, r_rest - 1);
    }
};
// @lc code=end

File: LeetCode/Java/22-backtracking-括号生成.java
/*
 * @lc app=leetcode.cn id=22 lang=java
 *
 * [22] 括号生成
 */

// @lc code=start
class Solution {
    private List<String> res;

    public List<String> generateParenthesis(int n) {
        res = new ArrayList<>();
        generate(0, 0, n, "");
        return res;
    }

    private void generate(int left, int right, int n, String s) {
        if (left == n && right == n) {
            res.add(s);
            return;
        }

        if (left < n)
            generate(left + 1, right, n, s + "(");

        if (left > right)
            generate(left, right + 1, n, s + ")");
    }
}
// @lc code=end
